src/effects/VST3/internal/x11/RunLoop.cpp

```diff
@@ -0,0 +1,147 @@
+/**********************************************************************
+
+  Audacity: A Digital Audio Editor
+
+  @file RunLoop.cpp
+
+  @author Vitaly Sverchinsky
+
+  @brief Part of Audacity VST3 module
+
+**********************************************************************/
+
+#include "RunLoop.h"
+
+using namespace internal::x11;
+
+struct RunLoop::EventHandler
+{  
+   Steinberg::IPtr<Steinberg::Linux::IEventHandler> handler;
+   GIOChannel* channel {nullptr};
+   guint id{0};
+
+   ~EventHandler()
+   {
+      if(id != 0)
+         g_source_remove(id);
+      if(channel != nullptr)
+         g_io_channel_unref(channel);
+   }
+};
+
+struct RunLoop::TimerHandler
+{
+   Steinberg::IPtr<Steinberg::Linux::ITimerHandler> handler;
+   guint id {0};
+
+   ~TimerHandler()
+   {
+      if(id != 0)
+         g_source_remove(id);
+   }
+};
+
+IMPLEMENT_FUNKNOWN_METHODS(RunLoop, Steinberg::Linux::IRunLoop, Steinberg::Linux::IRunLoop::iid);
+
+gboolean RunLoop::OnEvent(GIOChannel *source, GIOCondition condition, gpointer data)
+{
+   auto handler = static_cast<Steinberg::Linux::IEventHandler*>(data);
+   handler->onFDIsSet(g_io_channel_unix_get_fd(source));
+   return G_SOURCE_REMOVE;
+}
+
+gboolean RunLoop::OnTimer(gpointer data)
+{
+   auto handler = static_cast<Steinberg::Linux::ITimerHandler*>(data);
+   handler->onTimer();
+   return G_SOURCE_CONTINUE;
+}
+
+RunLoop::RunLoop(::Display* display)
+   : mDisplay(display)
+{
+   FUNKNOWN_CTOR;
+}
+
+RunLoop::~RunLoop()
+{
+   FUNKNOWN_DTOR;
+}
+
+Steinberg::tresult RunLoop::registerEventHandler(
+   Steinberg::Linux::IEventHandler* handler, 
+   Steinberg::Linux::FileDescriptor fd)
+{
+   using namespace Steinberg;
+
+   if(handler == nullptr)
+      return kInvalidArgument;
+    
+   if(XConnectionNumber(mDisplay) == fd)
+      return kResultOk;
+
+   auto eventHandler = std::make_unique<EventHandler>();
+   eventHandler->handler = handler;
+   
+   if(auto channel = g_io_channel_unix_new(fd))
+   {
+      eventHandler->channel = channel;
+      eventHandler->id = g_io_add_watch(
+         channel, 
+         (GIOCondition)(G_IO_IN | G_IO_OUT | G_IO_ERR | G_IO_HUP),
+         &RunLoop::OnEvent, 
+         handler
+      );
+      mEventHandlers.emplace_back(std::move(eventHandler));
```

Comment:
I think the destuctor of `eventHandler` destroys some resources too early, but you can fix this by giving the struct a move constructor that zeroes the id and channel pointer in the source of the move.

Reply:
Not sure that I understand. The type of `eventHandler` is `std::unique_ptr<>`, once the ownership is moved, `eventHandler` doesn't point to the object any more. Do I miss something here?

Reply:
You are right, it is a unique_ptr, and I overlooked that.

But implementing the move copy in EventHandler and TimerHandler, which act as resource handles, would be the right thing to do in those classes, considering what they do independently of how they are allocated: make sure that they are move-only and never are copied by mistake. It would allow you to avoid the unnecessary extra indirection of a unique_ptr.